### PR TITLE
Problem 2: Sort Characters by Frequency

### PR DESCRIPTION
**Pull request by hollowsunsets**

### CODE REVIEW

File: 2020-02-Spring/1_strings/README.md [hunk 1]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+#### Algorithm Overview
 
-#### Naive/Simple Solution
+1. Count the number of times each character occurs in the string
+2. Group the characters into strings with identical characters
+3. Sort the strings by length in ascending order
+4. Create the result string
 
-TODO :bug:
+There are many ways that you can do this. I've covered two solutions that use maps and arrays respectively, but you could also solve this using a priority queue.
```
Comment (huyn8) [inline 2020-02-Spring/1_strings/README.md:150]:
```suggestion
There are many ways that you can do this. I've covered two solutions that use arrays and maps respectively, but you could also solve this using a priority queue.
```
It said maps and arrays respectively but it went like solution 1: arrays, solution 2: maps so it was a little bit confusing.

File: 2020-02-Spring/1_strings/README.md [hunk 2]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+Make sure to use `StringBuilder` or similar language constructs so your solution does not have this extra cost.
+
+
+**Bucket Sort**
+
+There is a way you can solve this problem in `O(n)` time complexity, using bucket sort. Knowing to use bucket sort probably isn't expected for an interview since it's a somewhat obscure sorting algorithm, but still fun to know.
+
+#### Complexity Analysis
+
+**Solution 1**
```
Comment (huyn8) [inline 2020-02-Spring/1_strings/README.md:175]:
```suggestion
**Solution 1 (using arrays)**
```

File: 2020-02-Spring/1_strings/README.md [hunk 3]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+    * `O(n log n)` - Sorting the list of characters lexigraphically
+    * `O(n)` - Grouping characters into strings of identical characters
+    * `O(n)` - Sorting strings by length (if all of the characters are unique, in the worst case)
+    * `O(n)` - Building the result string
+
+* ***Space Complexity***: `O(n)`
+
+    * `O(n)` - Since Strings are immutable, we have to at least create a new string to return the result.
+
+**Solution 2**
```
Comment (huyn8) [inline 2020-02-Spring/1_strings/README.md:189]:
```suggestion
**Solution 2 (using maps)**
```
having a more descriptive title would make things more clear.

File: 2020-02-Spring/1_strings/README.md (hunk 1 above)
Comment (hollowsunsets) [reply 2020-02-Spring/1_strings/README.md:150]: Nice catch! :D Thanks

File: 2020-02-Spring/1_strings/README.md [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <!-- Don't remove -->
 <a name="top"/>
 
-# Topic Name TODO :bug:
+# Strings
```
Comment (huyn8) [inline 2020-02-Spring/1_strings/README.md:4]: don't need to do anything at this point but I changed the title name in the first PR. I can resolve any conflicts later.

File: 2020-02-Spring/1_strings/README.md [hunk 5]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+    * `O(n)` - Since Strings are immutable, we have to at least create a new string to return the result.
+
+**Solution 2 (using maps)**
+
+Suppose `n` is the length of the input string, and `k` is the number of unique characters in the string.
+
+* ***Time Complexity***: `O(n log n)`
+
+    * `O(n)` - Building the frequency map
+    * `O(k log k)` Sorting the frequency map by
```
Comment (etcadinfinitum) [inline 2020-02-Spring/1_strings/README.md:196]:
```suggestion
    * `O(k log k)` Sorting the frequency map by frequency of character occurrence (large to small)
```

File: 2020-02-Spring/1_strings/README.md [hunk 6]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+
+In the worst case, `k = n`, so we can say the final worst-case time complexity is still `O(n log n)`.
+
+* ***Space Complexity***: `O(n)`
+    * `O(n)` - Since Strings are immutable, we have to at least create a new string to return the result. Note that the frequency map is `O(k)`, but that is dominated by `O(n)`.
+
+##### Java
+
+<details>
+<summary>Click to see solution 1</summary>
```
Comment (etcadinfinitum) [inline 2020-02-Spring/1_strings/README.md:207]: Reiterate solution type, just for giggles
```suggestion
<summary>Click to see solution 1 (arrays)</summary>
```

File: 2020-02-Spring/1_strings/README.md [hunk 7]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+            result.append(s);
+        }
+        return result.toString();
+    }
+```
+</details>
+
+
+<details>
+<summary>Click to see solution 2</summary>
```
Comment (etcadinfinitum) [inline 2020-02-Spring/1_strings/README.md:252]: Reiterate solution type, just for giggles
```suggestion
<summary>Click to see solution 2 (maps)</summary>
```

File: 2020-02-Spring/1_strings/README.md [hunk 8]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+    }
+
+```
+
+</details>
+
+
+##### Python
+<details>
+<summary>Click to see solution 1</summary>
```
Comment (etcadinfinitum) [inline 2020-02-Spring/1_strings/README.md:298]:
```suggestion
<summary>Click to see solution 1 (arrays)</summary>
```

File: 2020-02-Spring/1_strings/README.md [hunk 9]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
+    return "".join(identical_groups)
+
+```
+
+
+</details>
+
+
+<details>
+<summary>Click to see solution 2</summary>
```
Comment (etcadinfinitum) [inline 2020-02-Spring/1_strings/README.md:336]:
```suggestion
<summary>Click to see solution 2 (maps)</summary>
```

File: 2020-02-Spring/1_strings/README.md [hunk 10]
```diff
@@ -119,21 +138,270 @@ Go to [Top](#top)
...
 
-#### Testing The Solutions OR Driver For Solution
+Remember to check if strings are immutable in your language of choice. If strings *are* immutable, you cannot modify them once they are created.
 
-TODO :bug:
+If you perform the following operation in Java:
+
+```java
+String websiteName = "This is a tomato site";
+websiteName += " for tomato friends";
```
[the remark this answers is not in the document]
Comment (hollowsunsets) [reply 2020-02-Spring/1_strings/README.md:161]: 🍅💓

File: 2020-02-Spring/1_strings/README.md (hunk 5 above)
Comment (hollowsunsets) [reply 2020-02-Spring/1_strings/README.md:196]: Thanks for catching this! 🌵

**Pull request merged**